### PR TITLE
ec2_scaling_policy - migrate to boto3

### PR DESCRIPTION
**Pull request by tremble**
##### SUMMARY

Based heavily on the work by @willthames : https://github.com/ansible/ansible/pull/26476

- Migrates ec2_scaling_policy to boto3
- Adds initial integration tests
- Adds support for StepScaling policies

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_scaling_policy

##### ADDITIONAL INFORMATION

Todo
- [x] Additional docs cleanup (suboptions)
- [x] Add AWSRetry handlers
- [x] AWS-Terminator policies
- [x] Changelog

Fixes: #152

### CODE REVIEW

**Comment by tremble** [comment]
blocked by https://github.com/mattclay/aws-terminator/pull/110

**Comment by actown** [comment]
This looks like it should cover my work in #152.

File: plugins/modules/ec2_scaling_policy.py [hunk 1]
```diff
@@ -6,178 +6,380 @@
...
+
+    if policy_type == 'StepScaling':
+        if not module.params['step_adjustments']:
+            module.fail_json(msg='step_adjustments is required when policy_type is StepScaling '
+                             'and state is present')
+        params['StepAdjustments'] = []
+        for step_adjustment in module.params['step_adjustments']:
+            step_adjust_params = dict(ScalingAdjustment=step_adjustment['scaling_adjustment'])
+            # Although empty bounds are allowed, not setting the 0 end of the bound seems
+            # to cause problems in the UI. We'll leave the +/- infinity end unset if not set
```
Comment (jillr) [inline plugins/modules/ec2_scaling_policy.py:280]: As best I can tell 0 and None both have the effect of setting infinity in the console, as does entirely commenting out lines 281, 282, 285, and 286 (the "= 0 if unset" bits).  Where did you run into issues?

I'm using https://github.com/ansible-collections/community.aws/blob/26eaf02cbec4ed2ed3a4681fea375af7522d83d7/tests/integration/targets/ec2_scaling_policy/tasks/main.yml#L145-L161 as a test followed by a pause so I can poke around the console.

**Review by jillr — COMMENTED** [review]
Thanks @tremble. Couple of questions, let me know if I'm just being dense on any of these.

File: plugins/modules/ec2_scaling_policy.py [hunk 2]
```diff
@@ -6,178 +6,380 @@
...
+    choices:
+      - Minimum
+      - Maximum
+      - Average
+    default: Average
+  step_adjustments:
+    type: list
+    description:
+      - list of dicts containing I(lower_bound), I(upper_bound) and I(scaling_adjustment)
+      - One item can not have a lower bound, and one item can not have an upper bound.
```
[the remark this answers is not in the document]
Comment (tremble) [reply plugins/modules/ec2_scaling_policy.py:85]: This is mostly Will's work, but yes.  Updated the wording, hopefully a little clearer now.

File: plugins/modules/ec2_scaling_policy.py (hunk 1 above)
Comment (tremble) [reply plugins/modules/ec2_scaling_policy.py:280]: This was Will's work, it's possible that Amazon have updated their UIs to fix the issue he was referring to.

Comment (tremble) [reply plugins/modules/ec2_scaling_policy.py]: I've removed the work-arounds and we can handle the issue as a bug (or raise a case with Amazon) if someone encounters the issue.

**Review by jillr — APPROVED** [review]

**Pull request merged**